### ISSUE TITLE
Rustc keeps suggesting std::iter::ZipImpl trait who has a "new" method.

### ISSUE THREAD

**Issue by crlf0710** [rust-lang/rust#45781]
This is not a real big deal, however rustc keeps suggesting "ZipImpl", which is a implementation detail and rarely is what people want.

Any chance to rename it somehow?

**Comment by Mark-Simulacrum** [comment]
Could you provide more context? A sample of the code and the error message suggesting `ZipImpl` would be helpful.

**Comment by ExpHP** [comment]
```rust
struct T;

fn main() {
    T::new();
}
```

```
   Compiling playground v0.0.1 (file:///playground)
error[E0599]: no function or associated item named `new` found for type `main::T` in the current scope
 --> src/main.rs:3:5
  |
3 |     T::new();
  |     ^^^^^^
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `new`, perhaps you need to implement it:
          candidate #1: `std::iter::ZipImpl`
```

**Comment by durka** [comment]
Maybe the suggestions should filter out `#[doc(hidden)]` items?

**Comment by ExpHP** [comment]
oddly enough, `ZipImpl` isn't just `#[doc(hidden)]`, but it is, in fact, [_private_.](https://github.com/rust-lang/rust/blob/dcbbfb6e807fdff9c9ba80073bb755f9d9d95e31/src/libcore/iter/mod.rs#L873-L883)

**Comment by zackmdavis** [comment]
Currently, we just check that at least one of the type and the trait is local, so indeed, this will false-positive for private traits:

https://github.com/rust-lang/rust/blob/565907fefb2a69c279b59bffa84c6d30129cafe9/src/librustc_typeck/check/method/suggest.rs#L445-L452

**Comment by estebank** [comment]
CC #25358, #26454.
